Introduce Mass Editor

MrBurrBurr opened the pull request:
## Description
Finally, ASF-ui users will be able to change the same specific bot-config values for all selected bots at once!

## Additional information
Closes #470

Abrynos commented:
I have a few suggestions for this PR:
- In case only one bot is selected: open the config editor modal for that bot
- I am not really a fan of the glow-effect when selecting bots. *In my opinion* it would be nicer if the bot image, name and status were darkened/greyed out when not selected and normal when selected
- When using the light theme the top text (e.g. "Select the bots oyu want to change config values for") might be a bit hard to read on some displays, as it is light grey on darker grey.
- I think the term "Editor" might be a bit confusing in the navbar on the left. Maybe change it for "Mass Editor"?
- I think it would be nicer to see the bots image, name and status again on the last page (just as on the first one, where we selected them) instead of just the bot names in some text. There is no need to modify the list there any more, but the visual would sure be nice ;)

Botan626 commented:
1. I think it should be called Mass Editor, not just Editor.
2. I think it should show current config values if one or more selected bots have the same config values for selected config properties.
3. I think it shouldn't offer to save config for bot, which already has selected config property.
4. The drop down menu for config properties looks bad in dark mode, it doesn't suit the dark mode, the scrollbar is invisible.
![image](https://user-images.githubusercontent.com/28785698/120920754-105a1f80-c6c9-11eb-8e08-4eb4302148c4.png)

Botan626 commented:
> * I am not really a fan of the glow-effect when selecting bots. _In my opinion_ it would be nicer if the bot image, name and status were darkened/greyed out when not selected and normal when selected

It is already grey for disabled bot.
![image](https://user-images.githubusercontent.com/28785698/120920878-b7d75200-c6c9-11eb-9405-3d7084908432.png)

Abrynos commented:
> It is already grey for disabled bot.

for a disabled one: yes. If you read my sentence again, you will see that I was talking about coloring based on selection and not on online-status. I didn't think of the top bar, that is colored, but the whole thing being a bit closer to the background color. including the name, the text "Disabled", the top color and the image. Maybe it would even be a good idea to make the color in the mass editor independant of the bots status and just use that bar for selection purposes. 🤔

Botan626 commented:
> > It is already grey for disabled bot.
> 
>  If you read my sentence again

All elements are already grey in dark mode without bot selection, even status line. You want to make them all black on black background?

> Maybe it would even be a good idea to make the color in the mass editor independant of the bots status and just use that bar for selection purposes.

What about consistency? That line shows bot's status.

Personally I like glowing effect very much.

MrBurrBurr commented:
Thank you very much for your feedback, @Abrynos and @Botan626! I would like to share my thoughts on some of the suggestions:

> In case only one bot is selected: open the config editor modal for that bot

I thought about this aswell but then decided against it as i felt it would break the whole point of the mass editor and create more follow "problems" then it solves.

> I am not really a fan of the glow-effect when selecting bots. _In my opinion_ it would be nicer if the bot image, name and status were darkened/greyed out when not selected and normal when selected

I can see that the glow effect is a hate-it-or-love-it kinda thing. It's somewhat hard for me to find a good solution for this since we are reusing the bot cards. I will try to find a solution thats not so intrusive as the glow. Maybe a simple green border will do. 

> When using the light theme the top text (e.g. "Select the bots oyu want to change config values for") might be a bit hard to read on some displays, as it is light grey on darker grey.

Yep good catch! I haven't looked at styling regarding the light theme so far (dark theme ftw)

> I think the term "Editor" might be a bit confusing in the navbar on the left. Maybe change it for "Mass Editor"?

You confirmed my thoughts, I changed it again.

> I think it would be nicer to see the bots image, name and status again on the last page (just as on the first one, where we selected them) instead of just the bot names in some text. There is no need to modify the list there any more, but the visual would sure be nice ;)

Alright, lets give it a try!

> I think it should be called Mass Editor, not just Editor, to reflect better its possibilities.

Yup, see answer above.

> I think it should show current config values, if one or more selected bots have the same config values for selected config properties.

Why? This just adds further complexity for no real advantage. Why would a user care what the current value is? He is defining a new value and it will be set for all selected bots.

> I think it shouldn't offer to save config properties for bots, which already have selected config properties.

I do not understand what you mean here.

> The drop down menu for config properties looks bad in dark mode, it doesn't suit the dark mode, the scrollbar is invisible.

Yes, this is still missing styling. No way I would have merged it like that :D

Botan626 commented:
> Why? This just adds further complexity for no real advantage. Why would a user care what the current value is? He is defining a new value and it will be set for all selected bots.

It's not about advantage, the user selects 2 bots, they both have `AutoSteamSaleEvent` set to true, he wants to modify this setting for both bots, he should see what the current value is for both bots here:
![image](https://user-images.githubusercontent.com/28785698/120933067-24217800-c701-11eb-8953-37f7a92b519f.png)

a) to avoid confusion, b) to remind him if he forgot, c) when modifying config from bot's card it's on too, d) you say 'Define the new config values', so why the new value was already selected without my actions?

> I do not understand what you mean here.

Bot 04 already has `AutoSteamSaleEvent` set to true, I can still choose `AutoSteamSaleEvent` to be true and save config. Why saving, if it's already true in config?
![image](https://user-images.githubusercontent.com/28785698/120933260-f426a480-c701-11eb-9ecb-6802ed39b113.png)

Abrynos commented:
I'm with @MrBurrBurr on this one. If someone wants to mass-change configurations they don't care about current values.

Botan626 commented:
> I'm with @MrBurrBurr on this one. If someone wants to mass-change configurations they don't care about current values.

What if half a year after last sale ended and new one is about to start, I want to set `AutoSteamSaleEvent` to true for all bots, but it's already true for all bots, but I forgot about it?

It should really show default values if current values are different for selected bots.

Botan626 commented:
> I will try to find a solution thats not so intrusive as the glow. Maybe a simple green border will do.

Maybe using the same effect as here:
![image](https://user-images.githubusercontent.com/28785698/120933859-bc6d2c00-c704-11eb-8d83-d67274171af3.png)
?

Botan626 commented:
I selected 2 bots, pressed F5, this is what I've got:
![image](https://user-images.githubusercontent.com/28785698/120934527-46b68f80-c707-11eb-9706-555f29e98955.png)

Botan626 commented:
> It should really show default values if current values are different for selected bots.

Or maybe it actually should always show default values, because selected bots could have different current values for selected config properties.

But then it should definitely disable saving functionality for bots, which already have the same current values for selected properties, like here:
![image](https://user-images.githubusercontent.com/28785698/120940376-a6bc2e80-c725-11eb-85e3-57386cc9a35d.png)

All these bots already have `AutoSteamSaleEvent` set to true.

On the last step with Save button bots could be divided in 2 parts, for the 1st part it could say 'Selected bots already have chosen values for selected properties!'.

Also, instead of this `Config values that will be changed:` isn't it  supposed to be `Config properties that will be changed:`?

And `Check the selected bots and the config values that will be saved` -> `Check selected bots, properties and values that will be saved`?

MrBurrBurr commented:
@Abrynos / @Botan626: Styling for light theme and dark theme should be done now 🎉 
Feel free to let me know what you think about it.

> Couple of consecutive F5 presses on Mass Editor page result in this:

Should be fixed with latest commit.

> But then it should definitely disable saving functionality for bots, which already have the same current values for selected properties

No need to try any further, this is out of scope. The scope for the Mass Editor is to set specific config values for selected bots.

> Also, instead of this 'Config values that will be changed:' isn't it supposed to be 'Config properties that will be changed:'?

No, because we do not change the config property, we change it's value.

Botan626 commented:
> No need to try any further, this is out of scope. The scope for the Mass Editor is to set specific config values for selected bots.

That's why it shouldn't save configs with unaltered values for selected properties.

> No, because we do not change the config properties, we change their values.

Config properties can have different values, we edit config properties and set up values for config properties.

Botan626 commented:
`Select the bots you want to change config values for` -> `Select the bots you want to edit config properties for`

Botan626 commented:
> Should be fixed with latest commit.

I still see error in console.
![image](https://user-images.githubusercontent.com/28785698/120943188-0589a400-c736-11eb-8689-a289c757349c.png)

Botan626 commented:
> Styling for light theme and dark theme should be done now 🎉
> Feel free to test it and let me know what you think about it.

I would say it looks great in dark mode now, but bad in light mode.
![image](https://user-images.githubusercontent.com/28785698/120943336-efc8ae80-c736-11eb-84ff-a486379872df.png)

MrBurrBurr commented:
> I still see error in console.

Lets try again: Should be fixed with latest commit.

> I would say it looks great in dark mode now, but bad in light mode.

Check latest commit. I think it is good enough. If you disagree then please provide some suggestions. No suggestion = no change.

Abrynos commented:
I like the new highlighting of selected bots. Tho I have a few more suggestions:
- The transition when you select them looks a bit laggy for me. Maybe decrease the time it takes. I tested a bit with my browser and I think it looks much better with `.1s` instead of `.3s`.
- There is no transition when DE-selecting a bot. The border just disappears instantly.
- When using the blue or the turquoise theme and selecting a bot that's currently "online" the two colors of the top bar and the border on the other edges make each other look bad. Maybe remove the indication of online-status by color completely and use the top portion of the thing for selection-indication as well. The information is in there as the text "Online" anyways.

I really like the whole feature. You did a great job there!

Botan626 commented:
> Lets try again: Should be fixed with latest commit.

is fixed

Botan626 commented:
> > No need to try any further, this is out of scope. The scope for the Mass Editor is to set specific config values for selected bots.
> 
> That's why it shouldn't save configs with unaltered values for selected properties.
> 
> > No, because we do not change the config properties, we change their values.
> 
> Config properties can have different values, we edit config properties and set up values for config properties.
> 
> `Values for selected properties will be changed:` sounds right.

> `Select the bots you want to change config values for` -> `Select the bots you want to edit config properties for`

What about all this @MrBurrBurr?

@JustArchi please tell us, what's your opinion on these matters?

JustArchi commented:
There is no reason why ASF-ui should not be smart about sending requests to ASF. ASF acts as a backend in this case, and UI *should* do whatever it can to ensure that no unnecessary requests will reach ASF. Telling ASF to save exact same config it gave you, provided you *know* that config in advance (and you do, you've just asked about it), is in fact unnecessary request.

^ ASF also has a potential here for improvement, as it could avoid saving the same content on the backend side. Problem is, I don't know the content of the file, and I'd need to read it first, compare, and then decide what to do. I already wasted I/O at this point, so I'm not saving any overhead as I don't know the current content. You do, you're actually modifying the config ASF has given to you, there is no logical argument against it, only "I'm too lazy to implement it" - which is totally fine, but not making it out of the scope. I mark such stuff with "PR-ok" and move on.

As of config properties, the exact definition is that you change values of config properties. From user standpoint however, he changes config properties. Config property is made out of the name, such as `AcceptGifts`, and value, such as `true`. Changing just one out of those two is enough to claim you, indeed, change config property. You could indeed go one level deeper into detail and state that you change *values* of config properties, but considering the fact that you don't have ambiguity here as you never change config properties *names*, I agree that it's better to just say "config properties" instead of adding (imho unnecessarily) "values" in addition to that.

JustArchi commented:
^ Be advised that the commit in ASF repo above does *not* address the enhancement in ASF-ui, ASF has its own optimization which can avoid the problem, but ASF-ui should *also* have its own optimization which will avoid the IPC API call.

Botan626 commented:
Thanks for thorough reply on both matters @JustArchi.

So Mass Editor could and should have logic and capabilities to tell users the properties they want to modify already have such values, and then disable saving functionality for that bots, if no other properties are going to have new values.

> From user standpoint however, he changes config properties.
> I agree that it's better to just say "config properties" instead of adding (imho unnecessarily) "values" in addition to that.

In addition to that, [Configuration](https://github.com/JustArchiNET/ArchiSteamFarm/wiki/Configuration) says that user can **edit** properties, I also think that it suits better, than 'change'.

I found the following phrases which can be improved @MrBurrBurr :

The 1st part of Mass Editor:
`Select the bots you want to change config values for` -> `Select the bots you want to edit config propereties for`

The 2nd part of Mass Editor:
`Select the config properties you want to change` -> `Select the config properties you want to edit`

The 3d part of Mass Editor:
`Change the config values for the config properties that you selected` -> `Edit the config properties you selected`

The last part of Mass Editor:
`Check the selected bots and config values` -> `Check the selected bots and config properties`
and
`Config values that will be changed:` -> `Config properties that will be edited:` or maybe it's better to say `The following config properties will be edited:`

Abrynos commented:
> tell users the properties they want to modify already have such values

I don't think it's necessary to tell users "5 of the 10 bots you wanna change already have those values". If a user wants to mass-edit bots, all he cares about is, that those bots have those values at the end of the day.

Botan626 commented:
> > tell users the properties they want to modify already have such values
> 
> I don't think it's necessary to tell users "5 of the 10 bots you wanna change already have those values". If a user wants to mass-edit bots, all he cares about is, that those bots have those values at the end of the day.

What if all selected properties already have same values for all selected bots?

> Telling ASF to save exact same config it gave you, provided you _know_ that config in advance (and you do, you've just asked about it), is in fact unnecessary request.

> You do, you're actually modifying the config ASF has given to you, there is no logical argument against it, only "I'm too lazy to implement it" - which is totally fine, but not making it out of the scope. I mark such stuff with "PR-ok" and move on.

With this implemented and Mass Editor disabling saving functionality in this case, you want to tell the user there's nothing to save, don't you?

Abrynos commented:
> What if all selected properties already have same values for all selected bots?

Do nothing and move on

Botan626 commented:
> Do nothing and move on.

How? What the user will see on the last part?

MrBurrBurr commented:
> The transition when you select them looks a bit laggy for me. Maybe decrease the time it takes. I tested a bit with my browser and I think it looks much better with `.1s` instead of `.3s`.

I changed the styling for it, de-selecting a bot should now also transition. Could you check if it is still laggy? If yes I will change the duration to '.1s'. Problem was that we did not set a border before so there was nothing to transition to. Only way to make this work (AFAIK) is to add a "invisible" border. But this creates other pixel-problems. So now I am starting to like your suggestion below more and more.. Tho Botan had a point where we should not change consistency. But if we changes the bot card to look differently from the "real" bot card I think we can make it work. We might as well remove the status text and change the online-status to be the selection status (as you suggested). Why would I want to know if the bot is online or farming or what ever? Hmm.. maybe someone has a good point against removing it. Other than that thank you for your feedback!

> I really like the whole feature. You did a great job there!

Thanks, @Abrynos  😎 

> In addition to that, Configuration says that user can edit properties, I also think that it suits better, than 'change'.

Good point in addition to the great response that @JustArchi gave. Thank you both.

> I found the following phrases which can be improved @MrBurrBurr :

Thanks, I added them all with a little exception to the last one. I do not like "edited" here. Seems kinda weird. I feel like "saved" fits better. Open for suggestions tho.

Botan626 commented:
> Thanks, I added them all with a little exception to the last one. I do not like "edited" here. Seems kinda weird. I feel like "saved" fits better. Open for suggestions tho.

Ok, I like it too.

Botan626 commented:
How about having Back button in the same position on the last past, as it is on the 2nd and 3d parts?
![image](https://user-images.githubusercontent.com/28785698/121024912-54265500-c7ad-11eb-87e1-1e411c0e9205.png)
![image](https://user-images.githubusercontent.com/28785698/121025075-79b35e80-c7ad-11eb-8e69-3aa34301c3e1.png)

This way it can be clicked several times without mouse movement.

MrBurrBurr commented:
> How about having Back button in the same position on the last past, as it is on the 2nd and 3d parts?

Creates more problems than it solves. Think about mobile view for example.

Botan626 commented:
> > How about having Back button in the same position on the last past, as it is on the 2nd and 3d parts?
> 
> Creates more problems than it solves. Think about mobile view for example.

Ok, it's just inconvenient to press it 3 times in order to return to the 1st part. Maybe there's another solution for this?

Also you need to add Mass Editor here:
![image](https://user-images.githubusercontent.com/28785698/121029390-36f38580-c7b1-11eb-8203-3408d1b0b686.png)

MrBurrBurr commented:
> Ok, it's just inconvenient to press it 3 times in order to return to the 1st part. Maybe there's another solution for this?

There is a solution for this. We could add some kind of "Steps" component that shows the current step we are in and user can navigate to different step by clicking on desired step.

> Also you need to add Mass Editor here:

Thanks, totally forgot about it!

Botan626 commented:
> There is a solution for this. We could add some kind of "Steps" component that shows the current step we are in and user can navigate to different step by clicking on desired step.

I don't know what you meant, but I'd like to see it,

Botan626 commented:
Should redirect to its 1st part (like F5 will do) after it successfully saves configs?
![image](https://user-images.githubusercontent.com/28785698/121033647-0ada0380-c7b5-11eb-9773-bc43ffc52fb5.png)

Nothing happens on screen besides notification and Save button can be pressed again.

MrBurrBurr commented:
> I don't know what you mean, but I'd like to see it.

I will implement this feature in a future PR. This one is already a big motherfucker and it works as expected currently.

> Should redirect to its 1st part (like F5 will do) after it successfully (does it check for it?) saves configs?

Not sure about this one.

Some user might like the fact that he can still go back after saving and do some changes and then re-save. Also I do not like the fact that we would "reset" the overview (meaning the selected bots and config properties) for the user.

Botan626 commented:
> I will implement this feature in a future PR. This one is already a big motherfucker and it works as expected currently.

Ok

> Not sure about this one.
> 
> Some user might like the fact that he can still go back after saving and do some changes and then re-save. Also I do not like the fact that we would "reset" the overview (meaning the selected bots and config properties) for the user.

I see the point, but maybe at lease disable Save button after configs were already successfully (does it check for it?) saved?

Botan626 commented:
> Thanks, totally forgot about it!

Did you want to order elements in this drop down menu the same way as they are on the left panel?
If yes, you want to change places for Log and Plugins.
![image](https://user-images.githubusercontent.com/28785698/121040935-fac52280-c7ba-11eb-8b06-c424b15371a7.png)

Botan626 commented:
I would replace this
`No config property found - Consider changing the search query`
with this
`No config property found, check your search query for errors`
![image](https://user-images.githubusercontent.com/28785698/121062944-6369ca80-c7ce-11eb-82bb-cdcea83a7e13.png)

MrBurrBurr commented:
> I see the point, but maybe at lease disable Save button after configs were already successfully (does it check for it?) saved?

Why? What problem does it solve? In my eyes it just creates a burden for some users.

> I would replace this 'No config property found - Consider changing the search query'

Again, why? I believe current one fits better.

Botan626 commented:
> Why? What problem does it solve? In my eyes it just creates a burden for some users.

To prevent false, unnecessary, excess config savings.

> Again, why? I believe current one fits better.

Too much of capital letters, dash is barely visible.

Botan626 commented:
> In my eyes it just creates a burden for some users.

What burden? The user pressed saved and then he can press save right away.

Botan626 commented:
![image](https://user-images.githubusercontent.com/28785698/121069194-e2aecc80-c7d5-11eb-8eed-55ba4e639141.png)

MrBurrBurr commented:
> What burden? The user pressed saved and then he can press save again right away.

Burden of going through whole wizard again. The "Save" button is already disabled while the saving process is in progress. It follows same logic as on ASF Config page. If the user is an idiot and clicks the save button 10 times even after getting the notification - not my problem.

Abrynos commented:
> If the user is an idiot and clicks the save button 10 times even after getting the notification - not my problem.

This is standard behaviour (saving even tho it is already saved). Every normal text editor (even MS Word) saves several times if you press save several times. The save button is only disabled while the saving-process is in progress. I fully agree that we should let the user be an idiot, if they want to be one.

Botan626 commented:
> Burden of going through whole wizard again.

User doesn't need to go back through wizard after he pressed Save button for the 1st time to choose new bots, properties and values?

> and will be again enabled once the saving process is finished.

Why? To save same config over and over again?
![image](https://user-images.githubusercontent.com/28785698/121071384-89946800-c7d8-11eb-8c9a-eaef00a1e831.png)

> It follows same logic as on ASF Config page.

This could be improved too.

> If the user is an idiot and clicks the save button 10 times even after getting the notification - not my problem.

Then why do you explain for an idiot user why this button is unclickable now?
![image](https://user-images.githubusercontent.com/28785698/121071987-543c4a00-c7d9-11eb-9345-24f4519cf029.png)

Botan626 commented:
> Every normal text editor (even MS Word) saves several times if you press save several times.

Wrong, Save button is disabled in Notepad++ if there's nothing to save.
![image](https://user-images.githubusercontent.com/28785698/121072884-7edad280-c7da-11eb-98dd-690dc378bdd4.png)

Aareksio commented on src/views/MassEditor.vue:
```diff
@@ -0,0 +1,311 @@
+    },
+    watch: {
+      async bots() {
+        // Needed because 'this.bots[0].name' is sometimes undefined
```
What exactly is _needed_ here? The `loadBotConfig` function is called and should throw an error, when `bots` array is empty. I do not see how the `watcherActive`, initially set to `true`, prevents this.

MrBurrBurr replied:
I have no idea but it does _some kind of magic_

Maybe something with the lifecycle? Watcher running after store is inizialized? I have no idea. But now that I refactored it to be its own component it _should_ be fixed and we do not need this. I will try it and report back.

MrBurrBurr replied:
Looks like it is fixed now. 👍

Botan626 commented:
Two more things:

1. You say stuff in singular form when it's suitable.
![image](https://user-images.githubusercontent.com/28785698/121073782-a8482e00-c7db-11eb-9286-73ce88640556.png)
![image](https://user-images.githubusercontent.com/28785698/121073744-98304e80-c7db-11eb-8f44-cf7cc04baa8a.png)
I think notification could also have the same logic, now it's only in plural form.
![image](https://user-images.githubusercontent.com/28785698/121074060-0248f380-c7dc-11eb-8fbb-81ccf14a1df7.png)

2. Pressing Esc on the 1st part when 1 or more bots are selected could unselect them, instead of sending user to the Bots page.
But then, ASF-ui sends user to the Bots page from any page, but Commands page, when Esc is pressed. What for? Any reason behind this?

Aareksio commented on src/views/MassEditor.vue:
```diff
@@ -0,0 +1,311 @@
+        // eslint-disable-next-line no-restricted-syntax
+        for (const [newProperty, newValue] of Object.entries(newConfig)) {
+          const found = Object.entries(oldConfig)
+            .find(([oldProperty, oldValue]) => oldProperty === newProperty && oldValue === newValue);
```
Why would you iterate over possibly all of the old properties instead of doing:
```js
const found = oldConfig[property] === newConfig[property]
```

MrBurrBurr replied:
To be honest I have no idea. Looking back at it your solution makes more sense. My excuse is that it was late at night and I was tired :D Thanks!

Aareksio reviewed (approved):
It's _alright_ :)

MrBurrBurr commented on src/components/MassEditor/Bots.vue:
```diff
@@ -0,0 +1,96 @@
+<template>
+  <div class="bots">
+    <div v-for="bot in bots" :key="bot.name" class="bot" :class="[`status--${bot.status}`, { selected: botIsSelected(bot), selectable }]" @click="select(bot)">
```
Well then the cursor would be shown in places where [it wouldn't make sense](https://user-images.githubusercontent.com/31552675/121088801-8c965500-c7e6-11eb-89e6-aad0dd0b536f.png).

MrBurrBurr replied:
My bad. I changed the styling to the following and it obviously works like a charm. Thanks, I learned something!

```scss
.selectable {
  > .bot {
    cursor: pointer;
  }
}
```

MrBurrBurr commented:
> Wrong, Save button is disabled in Notepad++ if there's nothing to save.

Feel free to send a PR if you want to help out those poor bastards that can't figure out that the config has been saved even after clicking the save button 10x times and seeing the notifications.

> I think notification could also have the same logic, now it's only in plural form.

Yep, I also wanted to change that but I do not like the current sentence. So feel free to suggest a nice sentence where we can use the plural forms.

> Pressing Esc on the 1st part when 1 or more bots are selected could unselect them, instead of sending user to the Bots page.

This sounds like something that no one would expect, or is it common to use the esc button to deselect something on a website?

> But then, ASF-ui sends user to the Bots page from any page, but Commands page, when Esc is pressed. What for? Any reason behind this?

This is default behavior of your browser. esc button acts as a 'reset' button. On commands page we are catching the key events since we focus the input on that page (so user can enter commands when he types on his keyboard).

Botan626 commented:
> Feel free to send a PR if you want to help out those poor bastards that can't figure out that the config has been saved even after clicking the save button 10x times and seeing the notifications.

Why so big objection just to make Save button disabled after 1st saving?

> Yep, I also wanted to change that but I do not like the current sentence. So feel free to suggest a nice sentence where we can use the plural forms.

Your notification is already in plural form only.

> This sounds like something that no one would expect, or is it common to use the esc button to deselect something on a website?

If I asked for it, then there is at least 1 person, who would expect it. Maybe there are others as well? It's intuitive to use Esc button to undo last actions, in this case to unselect bots.

> This is default behavior of your browser. esc button acts as a 'reset' button. On commands page we are catching the key events since we focus the input on that page (so user can enter commands when he types on his keyboard).

It is really a default behavior of my browser? Is it configurable? What does it reset? I don't want it to 'reset' anything. I open a youtube channel e.g., press Esc and it doesn't send me to youtube's home page.

Looks like a bug indeed.

MrBurrBurr commented:
> Why so big objection just to make Save button disabled after 1st saving?

Seems pointless for me so I do not want to be the one implementing it. But feel free to do so yourself. I will merge your PR if it works like you described.

> Your notification is already in plural form only.

Yes, and I do not like it in the current form.

> If I asked for it, then there is at least 1 person, who would expect it. Maybe there are others as well? It's intuitive to use Esc button to undo last actions, in this case to unselect bots.

So? I maybe expect the space button to unselect bots. Maybe there are others as well? Do you get my point? Besides that I would more expect ctrl+z to undo last action.

> It is really a default behavior of my browser? Is it configurable? What does it reset? I don't want it to 'reset' anything. I open a youtube channel e.g., press Esc and it doesn't send me to youtube's home page.

You see, I answer one question and you ask 3 more. If we keep going like this I won't be able to sleep tonight. :D Youtube is probably catching when someone is pressing the esc button without having any element in focus. We could also add something like this but this is something for a new issue then.

Botan626 commented:
> Feel free to send a PR if you want to help out those poor bastards that can't figure out that the config has been saved even after clicking the save button 10x times and seeing the notifications.

Moreover, as @JustArchi said:

> There is no reason why ASF-ui should not be smart about sending requests to ASF.

I think it's a great thing to male ASF-ui comprehensive.

MrBurrBurr commented:
> I think it's a great thing to male ASF-ui comprehensive.

So do I. And as I said before, feel free to send a PR.

Botan626 commented:
> Seems pointless for me

Strange.

> So? I maybe expect the space button to unselect bots. Maybe there are others as well? Do you get my point?

Your point was that:

> This sounds like something that no one would expect

Which is not true.

> We could also add something like this but this is something for a new issue then.

Add what? Your explanation about something `resetting` is at least questionable, as it's still unknown what is there resetting.
And error in console also suggests it's a bug in ASF-ui. Shall I open a bug issue on this matter?

Botan626 commented:
> Yes, and I do not like it in the current form.

Singular form:
`Config for selected bot has been saved!`

Plural form:
`Configs for selected bots have been saved!`

MrBurrBurr commented:
> > This sounds like something that no one would expect
> 
> Which is not true.

This is not how it works. If we go by your logic then the term itself will be useless. You can always find someone, somewhere in the world that would expect the most unexpected thing. But we are here talking about what the majority of people using ASF-ui (or a web browser in general) would expect. You can do a little experiment yourself: Go to ASF discord, offtopic channel and ask if someone would expect the esc button to deselect the bots. Let's see what the majority says.

> Add what? Your explanation about something `resetting` is at least questionable, as it's still unknown what is there been resetting.

If my explanation is questionable to you feel free to do your own research on that topic. 

> And an error in console also suggests it's a bug in ASF-ui. Shall I open a bug issue on this matter?

The error in the console is because you hit enter (which currently reloads the page) after already having been redirected to the page. Long story short. Open a new feature request where we catch if someone presses esc button and send "Event.preventDefault". This should fix it.

Botan626 commented:
> This is not how it works. If we go by your logic then the term itself will be useless. You can always find someone, somewhere in the world that would expect the most unexpected thing. But we are here talking about what the majority of people using ASF-ui (or a web browser in general) would expect.

I totally agree with that, but why you said that no one would expect that then? No need to answer that.

> You can do a little experiment yourself: Go to ASF discord, offtopic channel and ask if someone would expect the esc button to deselect the bots. Let's see what the majority says.

Maybe I will ask that question on Discord, need to think about it.

> The error in the console is because you hit enter (which currently reloads the page) after already having been redirected to the page.

I never pressed Enter. Where did you get that? Do you want me to make a gif with on-screen keyboard?

> Open a new feature request

Ok.

MrBurrBurr commented:
> I never pressed Enter. Where did you get that? Do you want me to make a gif with on-screen keyboard?

Typo. I ment the esc button.

Botan626 commented:
> > I never pressed Enter. Where did you get that? Do you want me to make a gif with on-screen keyboard?
> 
> Typo. I ment the esc button.

Ok, but I still need a clarification, opening Bots page by pressing Esc key on any page excl. Commands page is ok for you or not?

Botan626 commented:
> offtopic channel

Btw, why offtopic, if it's related to ASF development?

MrBurrBurr commented:
> Ok, but I still need a clarification, opening Bots page by pressing Esc key on any page excl. Commands page is ok for you or not?

No. This should be dealt with accordingly. As I already said: Feel free to open up feature request.

> Btw, why offtopic, if it's related to ASF development?

Better safe than sorry. Not sure everyone would define this as related to ASF development. For me it would be more of a general, off-topic question. But you do you.

Botan626 commented:
> Better safe than sorry. Not sure everyone would define this as related to ASF development. For me it would be more of a general, off-topic question. But you do you.

@JustArchi Is it ok if I will ask a question about deselecting bots by pressing Esc key on Mass Editor page on #asf-development channel?

JustArchi commented:
You can ask, but don't expect from me to organize a voting just because you want to get feedback from people.

I'd never expect ESC to do something meaningful on the website myself.

JustArchi commented:
But my opinion is irrelevant here, @MrBurrBurr decides what he wants to code and accept into **his** project, I can only offer feedback and suggestions, I'm not the guy you should be asking about it, because I'm not a maintainer of ASF-ui.

Botan626 commented:
> I'd never expect ESC to do something meaningful on the website myself.

This changed my position on this matter, so there will be no question. Also this gives more value for the feature request I will open re pressing Esc key.
